http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala

```diff
@@ -0,0 +1,205 @@
+package org.broadinstitute.dsde.workbench.leonardo.monitor
+
+import cats.effect.{Async, Concurrent, ContextShift, IO, Timer}
+import org.broadinstitute.dsde.workbench.google2.{Event, GoogleSubscriber}
+import fs2.{Pipe, Stream}
+import io.circe.{Decoder, DecodingFailure, Encoder}
+import org.broadinstitute.dsde.workbench.leonardo.MachineConfig
+import org.broadinstitute.dsde.workbench.leonardo.JsonCodec._
+import org.broadinstitute.dsde.workbench.leonardo.db.{clusterQuery, followupQuery, DbReference}
+import org.broadinstitute.dsde.workbench.leonardo.model.google.{ClusterStatus, MachineType}
+import org.broadinstitute.dsde.workbench.leonardo.util.ClusterHelper
+import _root_.io.chrisdavenport.log4cats.Logger
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.workbench.leonardo.model.LeoException
+import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterStatus.Stopped
+import org.broadinstitute.dsde.workbench.model.WorkbenchException
+
+import scala.concurrent.ExecutionContext
+
+case class ClusterNotFoundException(
+  clusterId: Long,
+  override val message: String =
+    "Could not process ClusterTransitionFinishedMessage because it was not found in the database"
+) extends LeoException
+
+class LeoPubsubMessageSubscriber[F[_]: Async: Timer: ContextShift: Logger: Concurrent](
+  subscriber: GoogleSubscriber[IO, LeoPubsubMessage],
+  clusterHelper: ClusterHelper,
+  dbRef: DbReference[IO]
+)(implicit executionContext: ExecutionContext)
+    extends LazyLogging {
+
+  def messageResponder(message: LeoPubsubMessage): IO[Unit] = {
+    val response = message match {
+      case msg @ StopUpdateMessage(_, _) =>
+        handleStopUpdateMessage(msg)
+      case msg @ ClusterTransitionFinishedMessage(_) =>
+        handleClusterTransitionFinished(msg)
+      case _ => IO.unit
+    }
+
+    //ensure we don't crash if the handler throws an exception
+    response.handleErrorWith(
+      e =>
+        IO(
+          logger
+            .error(s"Unable to process a message in received from pub/sub subscription in messageResponder: ${message}",
+                   e)
+        )
+    )
+
+    response
+  }
+
+  def messageHandler: Pipe[IO, Event[LeoPubsubMessage], Unit] = in => {
+    in.flatMap { event =>
+      for {
+        _ <- Stream
+          .eval(IO.pure(event.consumer.ack())) //we always ack first, as it could cause an endless loop of exceptions to do it after
```

Comment:
I decide to ack first here, otherwise Leo could concievably get stuck in a state where it keeps trying to process the same message and throwing errors. If it somehow gets into an undesirable state, I think it is better to simply flush the message out of the system, as manual intervention would be required otherwise.

Reply:
is this still true now that we're recovering from the errors? for some intermittent google errors, we probably should replay the messages right?

Reply:
Hmm.... I'm having trouble thinking of a scenario in which the order matters now, even if there are intermittent google errors, we would still ack as we handle these errors in the messageResponder